src/webui/api/torrentscontroller.cpp

```diff
@@ -1041,3 +1041,122 @@ void TorrentsController::categoriesAction()
 
     setResult(categories);
 }
+
+void TorrentsController::addTagsAction()
+{
+    checkParams({"hashes", "tags"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+    QJsonArray errors;
+
+    for (const QString &tag : tags) {
+        const QString tagTrimmed {tag.trimmed()};
+        if (BitTorrent::Session::instance()->isValidTag(tagTrimmed)) {
+            applyToTorrents(hashes, [&tagTrimmed, &errors](BitTorrent::TorrentHandle *const torrent)
+            {
+                if (!torrent->hasTag(tagTrimmed) && !torrent->addTag(tagTrimmed)) {
+                    errors.append(QJsonObject {
+                        {"tag", tagTrimmed},
+                        {"hash", static_cast<QString>(torrent->hash())},
+                        {"message", "Unable to add tag"}
+                    });
+                }
+            });
+        }
+        else {
```

Comment:
I'm fine with this format but @Chocobo1 prefer early returns (IMO, not always justified) so he can insist on reformatting it:
```c++
if (!BitTorrent::Session::instance()->isValidTag(tagTrimmed)) {
    // add error
    continue;
}

// applyToTorrents
```

Reply:
I do agree that we should return early.

> I'm fine with this format but @Chocobo1 prefer early returns (IMO, not always justified) so he can insist on reformatting it:

I'm not sure what the point of this comment was.